File: study/src/Patterns/proxy/Planet.java
package Patterns.proxy;

import Patterns.Display;
import Patterns.Drawable;

import java.awt.*;

public class Planet extends Drawable {
    private Color color;
    private int radius;
    SpaceShip dockedShip;


    public int getPx() {
        return x;
    }

    public int getPy() {
        return y;
    }


    public Planet(int x, int y, int radius, int r, int g, int b) {
        this.x = x;
        this.y = y;
        this.radius = radius;
        this.color = new Color(r, g, b);
    }

    @Override
    public void draw() {
        Graphics graphics = Display.getInstance().getGraphics();
        // draw the door
        graphics.setColor(color);
        graphics.fillArc(x - radius, y - radius, radius*2, radius*2, 0, 360);
    }

    public synchronized boolean dock(SpaceShip ship) {
        if(dockedShip == null) {
            dockedShip = ship;
            return true;
        }
        return false;
    }


    public synchronized void undock() {
        if(dockedShip != null) {
            try {
                dockedShip.undock();
            } catch (ShipHasNoDestinationException e) {
                throw new EverythingGoesBadException();
            }
            dockedShip = null;
        }
    }
}

File: study/src/kt/proxy_k/Planet.kt
package kt.proxy_k

import Patterns.Display
import Patterns.proxy.EverythingGoesBadException
import Patterns.proxy.ShipHasNoDestinationException
import Patterns.proxy.SpaceShip
import java.awt.*

open class Planet (
    x: Double,
    y: Double,
    private val radius: Double,
    val r: Int,
    val g: Int,
    val b: Int
    ): Drawable(x, y) {
    private val color = Color(r, g, b)
    var dockedShip: kt.proxy_k.SpaceShip? = null

    override fun draw() {
        // Display is java class - not overwritten
        val graphics = Display.getInstance().graphics
        graphics.color = color
        graphics.fillArc((x - radius).toInt(), (y - radius).toInt(), (radius * 2).toInt(), (radius * 2).toInt(), 0, 360)
    }

    @Synchronized
    fun dock(ship: kt.proxy_k.SpaceShip): Boolean {
        if (dockedShip == null) {
            dockedShip = ship
            return true
        }
        return false
    }

    @Synchronized
    fun undock() {
//        if (dockedShip != null) {
//            try {
//                    dockedShip!!.undock()
//                } catch (e: ShipHasNoDestinationException) {
//                    throw EverythingGoesBadException()
//                }
//        }

        dockedShip?.let{
                try {
                    it.undock()
                    dockedShip = null
                } catch (e: ShipHasNoDestinationException) {
                    throw EverythingGoesBadException()
                }
        }
    }


}